File: examples/kotlin/src/test/kotlin/adr/ToolsTest.kt
package adr

import kotlin.test.Test
import kotlin.test.assertEquals

class ToolsTest {

    private val ctx = Ctx(
        state = State.initial(listOf(Ticket("4118", "x", Priority.Normal, TicketStatus.Open))),
        actor = Actor.Agent,
    )

    @Test
    fun `a domain tool returns a payload and mutates nothing`() {
        val r = setPriority.run(SetPriorityInput("4118", Priority.High), ctx)
        assertEquals(ToolResult.PrioritySet("4118", Priority.High), r)
        // the context is unchanged — the tool was never handed a way to write it
        assertEquals(Priority.Normal, ctx.state.tickets.getValue("4118").priority)
    }

    @Test
    fun `a UI tool reads ctx read-only`() {
        assertEquals(ToolResult.Focused("4118", true), focusTicket.run(TicketInput("4118"), ctx))
        assertEquals(ToolResult.Focused("9999", false), focusTicket.run(TicketInput("9999"), ctx))
    }
}

File: examples/typescript/test/tools.test.ts
import { describe, it, expect } from "vitest";
import { initialState } from "../src/domain";
import { focusTicket, setPriority, type Ctx } from "../src/tools";

describe("tools are pure", () => {
  const ctx: Ctx = {
    state: initialState([{ id: "4118", body: "x", priority: "Normal", status: { kind: "Open" } }]),
    actor: "Agent",
  };

  it("a domain tool returns a payload and mutates nothing", () => {
    const r = setPriority.run({ ticket: "4118", level: "High" }, ctx);
    expect(r).toEqual({ kind: "PrioritySet", ticket: "4118", level: "High" });
    // the context is unchanged — the tool was never handed a way to write it
    expect(ctx.state.tickets.get("4118")!.priority).toBe("Normal");
  });

  it("a UI tool reads ctx read-only", () => {
    expect(focusTicket.run({ ticket: "4118" }, ctx)).toEqual({ kind: "Focused", ticket: "4118", known: true });
    expect(focusTicket.run({ ticket: "9999" }, ctx)).toEqual({ kind: "Focused", ticket: "9999", known: false });
  });
});
